zetta_utils/mazepa_addons/execution_tracker.py

```diff
@@ -107,3 +108,45 @@ def record_execution_run(execution_id: str) -> None:  # pragma: no cover
 
     with fsspec.open(info_path, "w") as f:
         json.dump(execution_run, f, indent=2)
+
+
+def read_execution_checkpoint(
+    filepath: str, ignore_prefix: Optional[Sequence[str]] = None
+) -> set[str]:  # pragma: no cover
+    """
+    Read completed IDs from checkpoint file.
+
+    `ignore_prefix`: skip IDs with matching prefix, e.g. `['flow-']`, default `None`.
+    """
+    if ignore_prefix is None:
+        ignore_prefix = []
+
+    logger.info(f"Reading execution checkpoint from {filepath}")
+    with fsspec.open(filepath, "r", compression=CHECKPOINT_COMPRESSION) as f:
+        completed_ids = json.load(f)
+    assert isinstance(completed_ids, list)
+
+    return set(
+        id_ for id_ in completed_ids if not any(id_.startswith(prefix) for prefix in ignore_prefix)
+    )
+
+
+def record_execution_checkpoint(
+    execution_id: str, ckpt_name: str, completed_ids: list[str]
+):  # pragma: no cover
+    """
+    Save completed tasks to checkpoint file
+    """
+    zetta_user = os.environ["ZETTA_USER"]
+    info_path = os.environ.get("EXECUTION_INFO_PATH", EXECUTION_INFO_PATH)
+    ckpt_path = os.path.join(
+        info_path, zetta_user, execution_id, f"{ckpt_name}.{CHECKPOINT_COMPRESSION}"
+    )
+
+    logger.info(f"Saving execution checkpoint to {ckpt_path}")
+    try:
+        with fsspec.open(ckpt_path, "w", compression=CHECKPOINT_COMPRESSION) as f:
+            json.dump(completed_ids, f, indent=2)
+
+    except Exception:  # pylint: disable=W0703
```

Comment:
I think this is a decision between _maybe_ saving time and _maybe_ saving cost:

gcsfs already has [retry logic for common exceptions](https://github.com/fsspec/gcsfs/blob/main/gcsfs/retry.py). The exceptions that end up here would most likely indicate something more permanent happened. E.g. the scheduler's write permissions to the checkpoint bucket got revoked, or the bucket itself was deleted. By keeping the deployment alive, you still have a chance to fix whatever caused this sudden change without losing any processing time (esp. if it really only was a temporary error that just exceeded gcfs retry count). If it is due to a broader issue that also affects the workers, then there is also no need to shut down the scheduler because of the failing checkpoint - the workers would take down the deployment, anyway.

How about a `raise_on_error` flag that will ensure the exception is raised and not just logged, and I call it once right at the start of the scheduler with strict behavior, to ensure the checkpointing setup has at least the _potential_ to work?

Reply:
`raise_on_checkpoint_failure` sounds like a good idea.

But regardless, we should be able narrow down the range of exceptions we catch. 
Something like ```(FileNotFoundError, IOError, gcsfs.utils.HttpError, requests.exceptions.ProxyError)```

Reply:
The argument about other threads possibly raising wild TimeoutErrors that are then caught by my Pokémon catch is convincing (and annoying)... I will fall back to "only" catching the errors that fsspec seems to be able to generate.
In addition, also add a flag that will cause checkpoint related exceptions to kill the deployment - in cases where gambling on a successful completion without any checkpoints is not desired. 🎲